drivers/hwinfo/hwinfo_ambiq.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2024 Ambiq
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <am_mcu_apollo.h>
+#include <zephyr/drivers/hwinfo.h>
+#include <string.h>
+#include <zephyr/sys/byteorder.h>
+
+struct ambiq_hwinfo {
+	uint32_t ui32ChipID0;
+	uint32_t ui32ChipID1;
+	uint32_t ui32TrimVer;
+};
+
+struct ambiq_hwinfo sAmbiqHwInfo = {0};
+
+ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+{
+	/* Contains the HAL hardware information about the device. */
+	am_hal_mcuctrl_device_t sMcuCtrlDevice;
+
+	am_hal_mram_info_read(1, AM_REG_INFO1_TRIM_REV_O / 4, 1, &sAmbiqHwInfo.ui32TrimVer);
+	am_hal_mcuctrl_info_get(AM_HAL_MCUCTRL_INFO_DEVICEID, &sMcuCtrlDevice);
+
+	sAmbiqHwInfo.ui32ChipID0 = sMcuCtrlDevice.ui32ChipID0;
+	sAmbiqHwInfo.ui32ChipID1 = sMcuCtrlDevice.ui32ChipID1;
+
+	if (length > sizeof(sAmbiqHwInfo)) {
+		length = sizeof(sAmbiqHwInfo);
+	}
+
+	memcpy(buffer, &sAmbiqHwInfo, length);
+
+	return length;
+}
+
+int z_impl_hwinfo_get_reset_cause(uint32_t *cause)
+{
+	uint32_t ui32Flags = 0;
+	uint32_t ui32ResetStatus = 0;
+	am_hal_reset_status_t sStatus = {0};
+
+	/* Print out reset status register upon entry */
+	am_hal_reset_status_get(&sStatus);
+	ui32ResetStatus = sStatus.eStatus;
+
+	/* EXTERNAL PIN */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_EXTERNAL) {
+		ui32Flags |= RESET_PIN;
+	}
+
+	/* POWER CYCLE */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_POR) {
+		ui32Flags |= RESET_POR;
+	}
+
+	/* BROWNOUT DETECTOR */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_BOD) {
+		ui32Flags |= RESET_BROWNOUT;
+	}
+
+	/* SOFTWARE POR */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_SWPOR) {
+		ui32Flags |= RESET_SOFTWARE;
+	}
+
+	/* SOFTWARE POI */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_SWPOI) {
+		ui32Flags |= RESET_SOFTWARE;
+	}
+
+	/* DEBUGGER */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_DEBUGGER) {
+		ui32Flags |= RESET_DEBUG;
+	}
+
+	/* WATCHDOG */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_WDT) {
+		ui32Flags |= RESET_WATCHDOG;
+	}
+
+	/* BOUNREG */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_BOUNREG) {
+		ui32Flags |= RESET_HARDWARE;
+	}
+
+	/* BOCORE */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_BOCORE) {
+		ui32Flags |= RESET_HARDWARE;
+	}
+
+	/* BOMEM */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_BOMEM) {
+		ui32Flags |= RESET_HARDWARE;
+	}
+
+	/* BOHPMEM */
+	if (ui32ResetStatus & AM_HAL_RESET_STATUS_BOHPMEM) {
+		ui32Flags |= RESET_HARDWARE;
+	}
+
+	*cause = ui32Flags;
+	return 0;
+}
+
+int z_impl_hwinfo_clear_reset_cause(void)
+{
+	/* SBL maintains the RSTGEN->STAT register in
```

Comment:
How to clear the status then? Is it cleared automatically after a reboot/reset?

Reply:
A user cannot clear the status. That is why it returns -ENOSYS as other hwinfo drivers do.

Reply:
```
> 
>   union {
>     __IOM uint32_t STAT;      /*!< (@ 0x0000885C) This register contains the status for brownout
>                                                   events and the causes for resets.
>                                                   NOTE 1: All bits in this register, including
>                                                   reserved bits, are writable. Therefore care
>                                                   should be taken not to write this register.
>                                                   NOTE 2: This register is only reset by POI
>                                                   not by HRESETn. Its contents are intended
>                                                   to survive all reset level except POI and
>                                                   full power cycles.                                         */
```
> 
>     struct {
>       __IOM uint32_t EXRSTAT    : 1;            /*!< [0..0] Reset was initiated by an External Reset.                          */
>       __IOM uint32_t PORSTAT    : 1;            /*!< [1..1] Reset was initiated by a Power-On Reset.                           */
>       __IOM uint32_t BORSTAT    : 1;            /*!< [2..2] Reset was initiated by a Brown-Out Reset.                          */
>       __IOM uint32_t SWRSTAT    : 1;            /*!< [3..3] Reset was a initiated by SW POR or AIRCR Reset.                    */
>       __IOM uint32_t POIRSTAT   : 1;            /*!< [4..4] Reset was a initiated by Software POI Reset.                       */
>       __IOM uint32_t DBGRSTAT   : 1;            /*!< [5..5] Reset was a initiated by Debugger Reset.                           */
>       __IOM uint32_t WDRSTAT    : 1;            /*!< [6..6] Reset was initiated by a Watchdog Timer Reset.                     */
>       __IOM uint32_t BOUSTAT    : 1;            /*!< [7..7] An Unregulated Supply Brownout Event occured.                      */
>       __IOM uint32_t BOCSTAT    : 1;            /*!< [8..8] VDDC Analog Brownout Event occured.                                */
>       __IOM uint32_t BOFSTAT    : 1;            /*!< [9..9] VDDF Analog Brownout Event occured.                                */
>       __IOM uint32_t BOSSTAT    : 1;            /*!< [10..10] VDDS Analog Brownout Event occured.                              */
>             uint32_t            : 21;
>     } STAT_b;
>   } ;